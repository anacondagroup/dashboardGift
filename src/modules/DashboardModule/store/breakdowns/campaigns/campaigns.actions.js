import {
  CAMPAIGNS_BREAKDOWN_LOAD_REQUEST,
  CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS,
  CAMPAIGNS_BREAKDOWN_LOAD_FAILS,
  CAMPAIGNS_DOWNLOAD_REPORT_REQUEST,
  CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS,
  CAMPAIGNS_DOWNLOAD_REPORT_FAIL,
} from './campaigns.types';

export const campaignsLoadRequest = options => ({
  type: CAMPAIGNS_BREAKDOWN_LOAD_REQUEST,
  payload: options,
});

export const campaignsLoadSuccess = breakdown => ({
  type: CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS,
  payload: breakdown,
});

export const campaignsLoadFail = error => ({
  type: CAMPAIGNS_BREAKDOWN_LOAD_FAILS,
  payload: error,
});

export const campaignsDownloadReportRequest = options => ({
  type: CAMPAIGNS_DOWNLOAD_REPORT_REQUEST,
  payload: options,
});
export const campaignsDownloadReportSuccess = () => ({
  type: CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS,
});
export const campaignsDownloadReportFail = () => ({
  type: CAMPAIGNS_DOWNLOAD_REPORT_FAIL,
});
