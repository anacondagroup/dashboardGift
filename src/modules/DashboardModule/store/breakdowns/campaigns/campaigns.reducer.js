import * as R from 'ramda';
import { renameKeys } from '@alycecom/utils';

import { initialBreakdownState } from '../../../helpers/breakdownSelectors.helpers';
import { setLoading, setLoaded } from '../../../../../helpers/lens.helpers';

import {
  CAMPAIGNS_BREAKDOWN_LOAD_REQUEST,
  CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS,
  CAMPAIGNS_BREAKDOWN_LOAD_FAILS,
  CAMPAIGNS_DOWNLOAD_REPORT_REQUEST,
  CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS,
  CAMPAIGNS_DOWNLOAD_REPORT_FAIL,
} from './campaigns.types';

export const initialState = {
  ...initialBreakdownState,
};

export const renameBreakdownKeys = renameKeys({
  gifts_sent: 'giftsSent',
  gifts_viewed: 'giftsViewed',
  gifts_accepted: 'giftsAccepted',
  meetings_booked: 'meetingsBooked',
  type: 'type',
});

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case CAMPAIGNS_BREAKDOWN_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };
    case CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS:
      return {
        ...state,
        ...R.compose(setLoading(false), setLoaded(true))(state),
        breakdown: action.payload.map(campaign => renameBreakdownKeys(campaign)),
      };
    case CAMPAIGNS_BREAKDOWN_LOAD_FAILS:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    case CAMPAIGNS_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS:
    case CAMPAIGNS_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    default:
      return state;
  }
};
