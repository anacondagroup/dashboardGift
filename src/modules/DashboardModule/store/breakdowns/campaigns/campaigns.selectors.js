import * as R from 'ramda';
import { updateSearch } from '@alycecom/modules';
import { pickNotEmpty, renameKeys } from '@alycecom/utils';

import { getBreakdownsState } from '../breakdowns.selectors';
import { makeGetTeamById } from '../../../../../store/teams/teams.selectors';
import { makeGetMemberById } from '../../members/members.selectors';
import { createSelectors } from '../../../helpers/breakdownSelectors.helpers';

const campaignsApiParams = R.pick(['dateRangeFrom', 'dateRangeTo', 'sort', 'sortDirection', 'search', 'teamId']);

export const [
  getCampaignsState,
  getCampaignsBreakdown,
  getCampaignsIsLoading,
  getCampaignsReportIsLoading,
  getCampaignsIsLoaded,
] = createSelectors(getBreakdownsState)('campaigns');

export const getApiCallQuery = R.compose(
  R.curryN(2, updateSearch)(''),
  renameKeys({
    teamId: 'team_id',
    sort: 'sort_column',
    sortDirection: 'sort_direction',
    search: 'search',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
  }),
  pickNotEmpty,
  campaignsApiParams,
);

export const getReportFileName = apiParams => state => {
  let team = null;
  let date = '';
  let member = null;

  if (apiParams.memberId) {
    const { fullName: memberName } = makeGetMemberById(apiParams.memberId)(state);
    member = `-${memberName}`;
  }

  if (apiParams.teamId) {
    const { name: teamName } = makeGetTeamById(apiParams.teamId)(state);
    team = `-${teamName}-Team`;
  }

  if (apiParams.dateRangeFrom) {
    date = `-${apiParams.dateRangeFrom}`;
  }

  if (apiParams.dateRangeTo) {
    date += `-${apiParams.dateRangeTo}`;
  }

  return R.replace(/\./g, '_', `Campaigns breakdown${team || ''}${member || ''}${date || ''}`).concat('.xlsx');
};

export const buildApiUrl = ({ teamId, memberId, ...rest }) => {
  if (teamId && memberId) {
    return `/enterprise/dashboard/breakdown/teams/${teamId}/members/${memberId}/campaigns?${getApiCallQuery(rest)}`;
  }

  if (teamId && !memberId) {
    return `/enterprise/dashboard/breakdown/teams/${teamId}/campaigns?${getApiCallQuery(rest)}`;
  }

  return `/enterprise/dashboard/breakdown/campaigns?${getApiCallQuery(rest)}`;
};

export const buildReportUrl = ({ teamId, memberId, ...rest }) => {
  if (teamId && !memberId) {
    return `/enterprise/report/teams/${teamId}/campaign_breakdown?${getApiCallQuery(rest)}`;
  }

  if (teamId && memberId) {
    return `/enterprise/report/teams/${teamId}/members/${memberId}/campaign_breakdown?${getApiCallQuery(rest)}`;
  }

  return `/enterprise/report/campaign_breakdown?${getApiCallQuery(rest)}`;
};
