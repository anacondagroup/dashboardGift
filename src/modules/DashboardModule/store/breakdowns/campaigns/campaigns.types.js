export const CAMPAIGNS_BREAKDOWN_LOAD_REQUEST = 'CAMPAIGNS_BREAKDOWN_LOAD_REQUEST';
export const CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS = 'CAMPAIGNS_BREAKDOWN_LOAD_SUCCESS';
export const CAMPAIGNS_BREAKDOWN_LOAD_FAILS = 'CAMPAIGNS_BREAKDOWN_LOAD_FAILS';

export const CAMPAIGNS_DOWNLOAD_REPORT_REQUEST = 'CAMPAIGNS_DOWNLOAD_REPORT_REQUEST';
export const CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS = 'CAMPAIGNS_DOWNLOAD_REPORT_SUCCESS';
export const CAMPAIGNS_DOWNLOAD_REPORT_FAIL = 'CAMPAIGNS_DOWNLOAD_REPORT_FAIL';
