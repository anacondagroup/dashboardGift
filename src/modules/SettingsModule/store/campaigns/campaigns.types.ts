import { CAMPAIGN_STATUS, CAMPAIGN_TYPES } from '../../../../constants/campaignSettings.constants';
import { IRowDataItem } from '../../../../components/Shared/CustomTable/CustomTable.types';

export interface ICampaignListItem extends IRowDataItem {
  id: number;
  name: string;
  status: CAMPAIGN_STATUS;
  type: CAMPAIGN_TYPES;
  invitationTypes: string[] | null;
  canEdit: boolean;
  team: {
    id: number;
    name: string;
  };
  recipientActions: Record<string, boolean> | null;
  countryIds: number[];
}
