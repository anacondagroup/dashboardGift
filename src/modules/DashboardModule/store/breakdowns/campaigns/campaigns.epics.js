import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, withLatestFrom, tap, debounce, takeUntil } from 'rxjs/operators';
import { timer } from 'rxjs';

import { CAMPAIGNS_BREAKDOWN_LOAD_REQUEST, CAMPAIGNS_DOWNLOAD_REPORT_REQUEST } from './campaigns.types';
import {
  campaignsLoadSuccess,
  campaignsLoadFail,
  campaignsDownloadReportSuccess,
  campaignsDownloadReportFail,
} from './campaigns.actions';
import { buildReportUrl, getReportFileName, buildApiUrl } from './campaigns.selectors';

export const loadCampaignsBreakdownEpic = (action$, state$, { apiService, messagesService: { errorHandler } }) =>
  action$.pipe(
    ofType(CAMPAIGNS_BREAKDOWN_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(({ payload }) => buildApiUrl(payload)),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => campaignsLoadSuccess(response.campaigns)),
        catchError(errorHandler({ callbacks: campaignsLoadFail })),
        takeUntil(action$.ofType(CAMPAIGNS_BREAKDOWN_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile, messagesService: { errorHandler } }) =>
  action$.pipe(
    ofType(CAMPAIGNS_DOWNLOAD_REPORT_REQUEST),
    map(({ payload }) => [buildReportUrl(payload), payload]),
    mergeMap(([url, payload]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => campaignsDownloadReportSuccess()),
        catchError(errorHandler({ callbacks: campaignsDownloadReportFail })),
      ),
    ),
  );

export const campaignsEpic = [loadCampaignsBreakdownEpic, downloadReportEpic];
