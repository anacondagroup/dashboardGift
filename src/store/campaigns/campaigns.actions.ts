import { createAction } from 'redux-act';

import { ICampaign } from './campaigns.types';

const PREFIX = 'DASHBOARD/CAMPAIGNS';

export const loadCampaignsRequest = createAction(`${PREFIX}/LOAD_CAMPAIGNS_REQUEST`);
export const loadCampaignsSuccess = createAction<ICampaign[]>(`${PREFIX}/LOAD_CAMPAIGNS_SUCCESS`);
export const loadCampaignsFail = createAction(`${PREFIX}/LOAD_CAMPAIGNS_FAIL`);
