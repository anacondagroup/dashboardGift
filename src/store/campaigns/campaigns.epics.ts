import { catchError, map, mapTo, mergeMap } from 'rxjs/operators';
import { Epic, ofType } from 'redux-observable';
import { handleError, handlers } from '@alycecom/services';

import { createCampaignSuccess } from '../../modules/SettingsModule/store/campaign/createCampaign/createCampaign.actions';
import { SWAG_SELECT_GENERATION_CODES_PROGRESS_FINISH } from '../../modules/SettingsModule/store/campaign/swagSelect/swagSelect.types';
import {
  campaignNameSaveSuccess,
  campaignSettingsOwnerSaveSuccess,
} from '../../modules/SettingsModule/store/campaign/general/general.actions';
import {
  expireProspectingCampaignById,
  unexpireProspectingCampaignById,
} from '../../modules/ProspectingCampaignModule/store/prospectingCampaign/prospectingCampaign.actions';

import { loadCampaignsSuccess, loadCampaignsFail, loadCampaignsRequest } from './campaigns.actions';
import { ICampaign } from './campaigns.types';

const loadCampaignsEpic: Epic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(loadCampaignsRequest),
    mergeMap(() =>
      apiService.get(`/enterprise/dashboard/campaigns`).pipe(
        map((response: { campaigns: ICampaign[] }) => loadCampaignsSuccess(response.campaigns)),
        catchError(handleError(handlers.handleAnyError(loadCampaignsFail))),
      ),
    ),
  );

const refreshListOnCreateCampaignEpic: Epic = action$ =>
  action$.pipe(ofType(createCampaignSuccess), mapTo(loadCampaignsRequest()));

const refreshListOnChangeNameEpic: Epic = action$ =>
  action$.pipe(ofType(campaignNameSaveSuccess), mapTo(loadCampaignsRequest()));

const refreshListOnChangeOwnerEpic: Epic = action$ =>
  action$.pipe(ofType(campaignSettingsOwnerSaveSuccess), mapTo(loadCampaignsRequest()));

const refreshListOnSwagCodesCreationFinishEpic: Epic = action$ =>
  action$.pipe(ofType(SWAG_SELECT_GENERATION_CODES_PROGRESS_FINISH), mapTo(loadCampaignsRequest()));

const refreshListOnProspectingCampaignActionsEpic: Epic = action$ =>
  action$.pipe(
    ofType(expireProspectingCampaignById.fulfilled, unexpireProspectingCampaignById.fulfilled),
    mapTo(loadCampaignsRequest()),
  );

export default [
  loadCampaignsEpic,
  refreshListOnCreateCampaignEpic,
  refreshListOnChangeNameEpic,
  refreshListOnChangeOwnerEpic,
  refreshListOnSwagCodesCreationFinishEpic,
  refreshListOnProspectingCampaignActionsEpic,
];
