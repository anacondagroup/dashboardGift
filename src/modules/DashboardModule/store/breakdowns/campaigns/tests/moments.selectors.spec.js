import { buildApiUrl, buildReportUrl, getReportFileName } from '../campaigns.selectors';
import { fullStateMock } from '../../../../../../mocks/fullState.mock';

describe('Campaigns selector', () => {
  describe('getReportFileName', () => {
    it('Should build name without params', () => {
      const actual = getReportFileName({})(fullStateMock);
      expect(actual).toMatchSnapshot();
    });

    it('Should build name with all params', () => {
      const actual = getReportFileName({
        teamId: '1',
        dateRangeFrom: '2019-03-22',
        dateRangeTo: '2019-03-24',
        memberId: '1210',
      })(fullStateMock);

      expect(actual).toMatchSnapshot();
    });
  });
  describe('buildApiUrl', () => {
    it('Should get api url with team and member ids', () => {
      const actual = buildApiUrl({
        teamId: 1,
        memberId: 4,
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });

    it('Should get api url with team id', () => {
      const actual = buildApiUrl({
        teamId: 1,
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });

    it('Should get api url', () => {
      const actual = buildApiUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });

  describe('buildReportUrl', () => {
    it('Should get api url with team and member ids', () => {
      const actual = buildReportUrl({
        teamId: 1,
        memberId: 4,
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });

    it('Should get api url with team id', () => {
      const actual = buildReportUrl({
        teamId: 1,
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });

    it('Should get api url', () => {
      const actual = buildReportUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });
});
