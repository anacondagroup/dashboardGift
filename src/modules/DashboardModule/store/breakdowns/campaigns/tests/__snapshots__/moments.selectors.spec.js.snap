// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campaigns selector buildApiUrl Should get api url 1`] = `"/enterprise/dashboard/breakdown/campaigns?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector buildApiUrl Should get api url with team and member ids 1`] = `"/enterprise/dashboard/breakdown/teams/1/members/4/campaigns?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector buildApiUrl Should get api url with team id 1`] = `"/enterprise/dashboard/breakdown/teams/1/campaigns?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector buildReportUrl Should get api url 1`] = `"/enterprise/report/campaign_breakdown?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector buildReportUrl Should get api url with team and member ids 1`] = `"/enterprise/report/teams/1/members/4/campaign_breakdown?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector buildReportUrl Should get api url with team id 1`] = `"/enterprise/report/teams/1/campaign_breakdown?search=asd&sort_column=name&sort_direction=asc"`;

exports[`Campaigns selector getReportFileName Should build name with all params 1`] = `"Campaigns breakdown-WEX: sales dep-Team-Shyanne Borer-2019-03-22-2019-03-24.xlsx"`;

exports[`Campaigns selector getReportFileName Should build name without params 1`] = `"Campaigns breakdown.xlsx"`;
